### ISSUE TITLE
CNN Transfer → Export + Upload to Hugging Face (Fusion-Compatible)

### ISSUE THREAD

**Issue by lukhsaankumar** [McMasterAI-Society/DeepFakeDetector#42]
## Issue Type

- [ ] **Model**: ML model bug, training issue, or architecture problem
- [ ] **Data**: Dataset issue, preprocessing bug, or data pipeline problem  
- [ ] **Web**: Frontend bug or UI issue in the Next.js dashboard
- [ ] **API**: Backend API bug or FastAPI endpoint issue
- [ ] **Research**: Research question or experimental feature request
- [ ] **Documentation**: Documentation bug or improvement needed
- [ ] **Bug**: General bug fix needed
- [x] **Enhancement**: New feature or improvement request

## Description
Standardize CNN Transfer model export artifacts AND publish them to Hugging Face Hub so the fusion model + backend can load them remotely and deterministically. 

This ticket updates the workflow from:
- **Previous**: `save artifacts to models/cnn_transfer/ locally`
- **New**: `export artifacts → upload to Hugging Face repo → backend downloads from Hub`

**Hugging Face repo**: DeepFakeDetector/cnn-transfer (https://huggingface.co/DeepFakeDetector/cnn-transfer)

## Acceptance Criteria
- [ ] A Hugging Face **Write token** is created and stored locally as an environment variable (DO NOT commit token)
- [ ] `cnn_transfer_baseline.ipynb` includes an **Export + Upload** section at the end
- [ ] Artifacts exported in a consistent folder layout:
  - `model.pth` (PyTorch state_dict)
  - `config.json` (architecture + threshold + misc)
  - `preprocess.json` (input size, normalization, crop behavior)
  - `label_map.json` (0 real, 1 fake)
  - `README.md` (short model card including label convention)
- [ ] Artifacts successfully uploaded to HF repo
- [ ] A quick local reload test is included: download snapshot → load weights → run 1 inference

### Hugging Face (source of truth)
Uploaded to: https://huggingface.co/DeepFakeDetector/cnn-transfer

Must contain:
- `model.pth`
- `config.json`
- `preprocess.json`
- `label_map.json`
- `README.md`

### Notebook update
`cnn_transfer_baseline.ipynb` has final cells that:
1. Export artifacts to a folder
2. Upload that folder to HF

### Label Convention (CRITICAL)
**Label convention MUST be fixed**:
- `0 = real`
- `1 = fake`

### Configuration
- Store **threshold** in `config.json` (default 0.50 unless tuned)

### Downstream Usage
This repo is later consumed by:
- `fusion` training script
- backend `model_registry` download/load pipeline

---

### 1) Create Hugging Face Write Token
Create a write token under your Hugging Face account settings and export locally:

**Environment variable (recommended):**
```bash
export HF_TOKEN_WRITE=...
```

**Never commit this token to git.**

### 2) Notebook Cell Snippets to Add
Add these cells to the end of `cnn_transfer_baseline.ipynb`:

#### Cell A — Export artifacts (weights + configs)
```python
import os, json, torch

# ====== CONFIG YOU MAY EDIT ======
EXPORT_DIR = "export/cnn-transfer"
os.makedirs(EXPORT_DIR, exist_ok=True)

# Save weights (state_dict)
torch.save(model.state_dict(), os.path.join(EXPORT_DIR, "model.pth"))

# Save preprocess details (match whatever you used in training!)
preprocess = {
    "input_size": 224,
    "center_crop": True,
    "normalize": {
        "mean": [0.485, 0.456, 0.406],
        "std": [0.229, 0.224, 0.225]
    }
}

# Save label map convention
label_map = {"0": "real", "1": "fake"}

# Save model config (fill in with your real backbone name, etc.)
config = {
    "name": "cnn-transfer",
    "framework": "pytorch",
    "arch": "resnet50",  # change if different
    "num_classes": 2,
    "threshold": 0.50,
    "labels": label_map,
    "notes": "CNN transfer baseline for deepfake detection"
}

with open(os.path.join(EXPORT_DIR, "preprocess.json"), "w") as f:
    json.dump(preprocess, f, indent=2)

with open(os.path.join(EXPORT_DIR, "label_map.json"), "w") as f:
    json.dump(label_map, f, indent=2)

with open(os.path.join(EXPORT_DIR, "config.json"), "w") as f:
    json.dump(config, f, indent=2)

# Minimal README model card
readme = """---
license: apache-2.0
tags:
  - image-classification
  - deepfake-detection
---

# CNN Transfer (DeepFakeDetector)

Binary classifier for deepfake / AI-generated image detection.

## Labels
- 0 = real
- 1 = fake

## Output
- prob_fake in [0,1] (sigmoid)
"""

with open(os.path.join(EXPORT_DIR, "README.md"), "w") as f:
    f.write(readme)

print("Exported CNN Transfer artifacts to:", EXPORT_DIR)
```

#### Cell B — Upload folder to Hugging Face Hub
```python
import os
from huggingface_hub import HfApi

# HF repo to upload to:
REPO_ID = "DeepFakeDetector/cnn-transfer"

# Token from environment variable (recommended)
HF_TOKEN = os.environ.get("HF_TOKEN_WRITE")
assert HF_TOKEN, "Missing HF_TOKEN_WRITE env var. Set it before running this cell."

api = HfApi(token=HF_TOKEN)

# Create repo if it doesn't exist (safe)
api.create_repo(repo_id=REPO_ID, repo_type="model", exist_ok=True)

# Upload everything in the export folder
api.upload_folder(
    folder_path=EXPORT_DIR,
    repo_id=REPO_ID,
    repo_type="model",
    commit_message="Upload CNN transfer artifacts (model + preprocess + config)"
)

print(f"Uploaded to https://huggingface.co/{REPO_ID}")
```

#### Cell C — (Optional but recommended) Sanity test: download + reload
```python
import torch
from huggingface_hub import snapshot_download

local_repo = snapshot_download(repo_id=REPO_ID, repo_type="model")

# Rebuild your architecture exactly as in training, then:
state = torch.load(os.path.join(local_repo, "model.pth"), map_location="cpu")
model.load_state_dict(state)
model.eval()

print("Reloaded model weights from HF snapshot:", local_repo)
```

## Definition of Done
- [x] HF repo contains the 5 required files
- [x] Notebook can export + upload in one run
- [x] Reload test passes
- [x] Labels & preprocess are clearly documented

**Comment by coolguy-stack** [comment]
[cnn_transfer_baseline_test.ipynb](https://github.com/user-attachments/files/25353249/cnn_transfer_baseline_test.ipynb)

HuggingFace repo: https://huggingface.co/DeepFakeDetector/cnn-transfer/tree/main
